### ISSUE TITLE
Fn + F5 / F6 / F7 (previous song, pause, next song) buttons have stopped working

### ISSUE THREAD

**Issue by krevs** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#3920]
**Checklist:**

* [ ] I have checked that there are no issues with similar or the same content

  **YOU SHOULD CHECK CLOSED ISSUES AS WELL**
* [ ] I have checked the FAQ (https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/wiki/FAQ) and the answer I am looking for is not there
* [ ] I have double checked and can reproduce the issue

**OS:** Windows

**GPMDP Version:** 4.7.1

**Issue Descriptions:** Fn + F5 / F6 / F7 (previous song, pause, next song) buttons have stopped working

**Steps to Reproduce:**

**Comment by cruzzer** [comment]
I'm seeing a similar issue on Windows and MacOS, where the media keys no longer work.

**Comment by xc-racer** [comment]
Similar issue as of this morning. 
Play/Pause, Previous Track & Next Track keys on my Logitech K330 keyboard have stopped working. 

OS - Windows 10 Home version 1909
GPMDP Version - 4.7.1

**Comment by stick1300** [comment]
Yup, noticed this late yesterday. I've got the previous/next/pause commands mapped to custom keys (SHIFT+CTRL+I/O/P). 

OS - Windows 10 Pro 19041.572 
GPMDP version - 4.7.1

**Comment by tracep** [comment]
I can confirm it's happening as well. None of my hotkeys that have worked for years have stopped working. I used a hotkey detective to confirm that GPMDP is still the only app that they are mapped to.  I also wiped out my AppData for GPMDP to reset it and no luck.

OS - Windows 10 Pro 20H2 (19042.804)
GPMDP - 4.7.1

**Comment by cfsmp3** [comment]
Actually it's not the keys. I've spent a few hours tracking this down (linux) trying to figure out if there was any program hijacking the keys, but no.

If you use the same options on the menu they don't work either. Strange, as "Show" for example works as expected, as well as the rest of the menu (quit, help, etc).

I suspect the problem is not with GPMDP but some change from Google breaking things.

**Comment by weaseldotro** [comment]
Pause, next and previous track keys have stopped working when using this player on my machine with Ubuntu 20.04.

**Comment by bruno-brant** [comment]
Same here - I configure personal hotkeys like CTRL+ALT+C and CTRL+ALT+V and they simply stopped working, out of the blue. Also my Logitech headset media keys. 

I restarted Windows a couple of times to see if GPMDP was able to register the keys and tried to check if any other program was hijacking those keys but I concluded that this is something with GPMDP itself. I now certain of it because of this thread - people on different systems having similar trouble.

Is there a log that I could look at? Although, I'd say that GPMDP is oblivious to the issue, since usually it complains right after I set the key, and this isn't the case.

**Comment by bruno-brant** [comment]
Reflecting on the issue (without looking at the code) I expect that something changed on YT side of things because:

- It's started without any changes to GPMDP itself.
- It's affecting people on all kind of OSes.
- GPMDP isn't complaining when registering keys, it's just not doing what it's supposed to do once we actuate those keys.

UPDATE: Right clicking the taskbar icon and selecting Play/Pause doesn't work either, which confirms my assumptions.

![image](https://user-images.githubusercontent.com/1890900/109971495-1826b400-7cd5-11eb-8598-84d9453777c5.png)

**Comment by bruno-brant** [comment]
I think I tracked down the issue. The culprit is actually a upstream project, (ytmusic.js)[gmusic-utils/ytmusic.js]. GPMDP relies on it to control playback on YT music. I believe its selector might be failing o click the button. If I'm not mistaken, the selectors are [here](https://github.com/gmusic-utils/ytmusic.js/blob/3d7e26405a31994519ae908d9be50fe1983895e2/src/constants/selectors.js). For play/pause for instance, they are:

```js
  playPause: '.left-controls .play-pause-button',
```
And while I can't attest to what was before, this selector returns null:

![image](https://user-images.githubusercontent.com/1890900/109975136-119a3b80-7cd9-11eb-9011-1bdd95ea9833.png)

While using just `.play-pause-button'` might be enough:

![image](https://user-images.githubusercontent.com/1890900/109975416-5faf3f00-7cd9-11eb-8487-3ceea81aa2ef.png)

(this last snippet will start/stop playing in YT Music)

**Comment by jostrander** [comment]
Appears to be related to track progress, investigating further.

**Comment by FinchPowers** [comment]
In the meantime, it seems like the command keys are working in the browser version. (I've tried with Firefox.)

**Comment by xc-racer** [comment]
Command keys are working in Microsoft Edge, but NOT in Chrome (for me).

**Comment by HudsonAfonso** [comment]
<img width="900" alt="20210304_050149-CapturFiles" src="https://user-images.githubusercontent.com/952447/110023168-9900a280-7d0b-11eb-8342-655b7fa86820.png">

**Comment by cruzzer** [comment]
Does this mean that the issue has been fixed upstream and now needs to be compiled into a new release of GPMDP?

**Comment by agzam** [comment]
Any news on this issue? This is so infuriatingly frustrating. I have paid account for Google Music for which Google has never built a native player, nor ever tried to support this one, or at least ever attempted to make sure not to break things. So many times I had to deal with their temper-tantrums. And so many times I would get frustrated, yet keep using Google Music, which is now YouTube Music.

Honestly, I feel guilty for using anything from Google these days, and even more so for paying for its services.
Google not even trying to hide it anymore, they are not on the trajectory to become evil - they already are evil.

I feel so depressed and helpless sometimes for not being able to say: "Go to Hell, Google! Go to Hell with all your shit!". Using anything of Google these days feels like smoking cigarettes in front of your kids. Gross, stupid and inconsiderate. I wouldn't even get surprised if at some point people organize "support groups", similar to AA, where they'd be helping each other to stay off Google and its services. "Wohoo! Today I got another star for not using Gmail for six months straight. I'm doing it folks! I'm free!"...

**Comment by cfsmp3** [comment]
> Any news on this issue? This is so infuriatingly frustrating. I've got a paid account for Google Music, for which Google has 

You realize that is project has nothing to do with Google, right :-)
It seems to be an abandoned project, so it's up to the community to fork and maintain it.
I tried that fix and it didn't work for me, but maybe I was having a bad idea and didn't do it right, or I had some other problem as well.

What specific OS are you using?

**Comment by agzam** [comment]
> You realize that is project has nothing to do with Google, right :-)

Oh, of course I am. I apologize for that it sounded like I'm complaining about the project. No. I completely understand that this issue is upstream and like it happened many times before - it is due to Google's BS.

I am not frustrated with this project. I love it.

**Comment by DanRxn** [comment]
For whatever it's worth, the menu command, `Edit > Previous Page`, is not working for me, to go back to the previous screen. Neither is the keyboard shortcut for the same command: 

`Cmd` + `[`

Feels like this has broken basically at the same time I noticed the media keys (play, pause, skip, etc.) not working.  Just mentioning in case that helps anyone have a better sense of where this may be breaking.  Even if not helpful, hopefully this will get fixed along with the media keys...

🤞😬

**Comment by bruno-brant** [comment]
@DanRxn same here. I was about to post about it - I use mouse keys for back/forward and they stopped working. Alt+→ or Alt+← also don't work. Probably it's all related.

**Comment by trobrock** [comment]
Does anyone know what needs to be done in the code to get this working again? It crazy annoying and I'm happy to dedicate some time today to fixing it if I could be pointed in the right direction (if there is even anything we can do within this project).

**Comment by cruzzer** [comment]
> 
> 
> Does anyone know what needs to be done in the code to get this working again? It crazy annoying and I'm happy to dedicate some time today to fixing it if I could be pointed in the right direction (if there is even anything we can do within this project).

There appears to have been a google side change which the player hooked into. In @bruno-brant [comment](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/3920#issuecomment-790641150), he mentions an upstream project which is used by GPMDP to control ytmusic.

In a linked ytmusic.js [issue](https://github.com/gmusic-utils/ytmusic.js/issues/24), they mentioned that they've fixed an issue on their side and comment that GPMDP will need to pull their change. You might just need to build GPMDP with their changes.

**Comment by SlashmanX** [comment]
> Does anyone know what needs to be done in the code to get this working again? It crazy annoying and I'm happy to dedicate some time today to fixing it if I could be pointed in the right direction (if there is even anything we can do within this project).

#3922 just needs to get merged

**Comment by bruno-brant** [comment]
Everyone coming here for this fix: you can get a pre-release from AppVeyor (either go here https://ci.appveyor.com/project/MarshallOfSound/google-play-music-desktop-player-unofficial/builds/38077585/artifacts or follow the link in the PR #3922). 

This is a work around while said PR isn't merged.

**Comment by cfsmp3** [comment]
Ultimately I switched to Youtube Music Desktop which is maintained and I'd say pretty much the same thing (and it works).

**Comment by agzam** [comment]
@cfsmp3 Nope. This feature - Multimedia keys (Previous, Next, Pause, etc.) currently broken in ytmdesktop-youtube-music app as well.

**Comment by YellowGTO** [comment]
> 
> 
> Ultimately I switched to Youtube Music Desktop which is maintained and I'd say pretty much the same thing (and it works).

Same. For anyone else, if you navigate to music.youtube.com with chrome, near the chromecast button it will ask you to install the desktop app.

**Comment by cfsmp3** [comment]
On Tue, Mar 16, 2021 at 3:54 PM Ag Ibragimov ***@***.***>
wrote:

> @cfsmp3 <https://github.com/cfsmp3> Nope. This feature - Multimedia keys
> (Previous, Next, Pause, etc.) currently broken in ytmdesktop-youtube-music
> app as well.
>
I'm using them just fine.

>

**Comment by agzam** [comment]
> if you navigate to music.youtube.com with chrome, near the chromecast button it will ask you to install the desktop app.

That thing has never worked for me. It just hangs. It hangs both itself and the browser. Tried with Chrome and Brave - same results. Probably has something to do with OS X Big Sur settings. Fuck you Google! and Fuck you Apple! (you ain't no better)

**Comment by tconley** [comment]
i had to switch to youtube desktop music to get my shortcut keys working :(

**Comment by martinduparc** [comment]
Here is a macOS build with the updated ytmusic.js
https://www.dropbox.com/s/8wg2lrlxjnwd9r0/Google%20Play%20Music%20Desktop%20Player.zip

**Comment by cruzzer** [comment]
> 
> 
> Everyone coming here for this fix: you can get a pre-release from AppVeyor (either go here https://ci.appveyor.com/project/MarshallOfSound/google-play-music-desktop-player-unofficial/builds/38077585/artifacts or follow the link in the PR #3922).
> 
> This is a work around while said PR isn't merged.

Can confirm that this worked for me on Windows 10. Media keys are working again. Thank you.

Can also confirm that the program referenced by @cfsmp3, Youtube Music Desktop worked for me. I had not heard of it before. As probably most, I searched for a desktop player when I first got Google Music and have stuck with GPMDP ever since.

**Comment by azor738** [comment]
> 
> 
> Everyone coming here for this fix: you can get a pre-release from AppVeyor (either go here https://ci.appveyor.com/project/MarshallOfSound/google-play-music-desktop-player-unofficial/builds/38077585/artifacts or follow the link in the PR #3922).
> 
> This is a work around while said PR isn't merged.

This fixed the issue for me, thanks so much!!!! ( ﾟДﾟ)b

**Comment by DanDobrick** [comment]
> Here is a macOS build with the updated ytmusic.js
> https://www.dropbox.com/s/8wg2lrlxjnwd9r0/Google%20Play%20Music%20Desktop%20Player.zip

This build works perfectly, thanks so much!

**Comment by cruzzer** [comment]
> 
> On Tue, Mar 16, 2021 at 3:54 PM Ag Ibragimov ***@***.***> wrote: @cfsmp3 <https://github.com/cfsmp3> Nope. This feature - Multimedia keys (Previous, Next, Pause, etc.) currently broken in ytmdesktop-youtube-music app as well.
> I'm using them just fine.
> […](#)

I had confirmed that it worked for me on Windows 10.
Today I tested and could not get it to work with MacOS. I tried various versions. Maybe someone had more luck with MacOS.

**Comment by agzam** [comment]
> This build works perfectly, thanks so much!

Fuck, yeah! I love you guys! Thank you for saving me from this misery. It's been extremely annoying without it.

**Comment by jgspratt** [comment]
That dropbox link is no longer a .zip.  Can you please upload another?  I need a copy.

Edit: Ignore me.  You can't right-click -> download.  Left click it and follow the menu.

**Comment by jgspratt** [comment]
I can confirm the dropbox build above works.  I had a session open so long that it only today exhibited this issue.

**Comment by lbarnes-cs** [comment]
> Here is a macOS build with the updated ytmusic.js
> https://www.dropbox.com/s/8wg2lrlxjnwd9r0/Google%20Play%20Music%20Desktop%20Player.zip

Tested on MacOS (Big Sur 11.2.3) and confirmed that it works with the media keys. Cannot express my gratitude enough. Thank you!!

**Comment by elfbread420** [comment]
> Everyone coming here for this fix: you can get a pre-release from AppVeyor (either go here https://ci.appveyor.com/project/MarshallOfSound/google-play-music-desktop-player-unofficial/builds/38077585/artifacts or follow the link in the PR #3922).
> 
> This is a work around while said PR isn't merged.

This did work for me, but not on my new computer. The windows 10 media overlay shows up for me, but with no Track playing. The Youtube Music app (installed through Chrome) works now, but unfortunately you can't have your Chrome and Youtube Music app logged into separate accounts. Very annoying.

Any suggestions beyond just using a different browser for YouTube (what I'm doing now)?

Edit: Spoke too soon, only pausing works. Even unpausing doesn't work.

**Comment by agarzola** [comment]
The build shared by @martinduparc worked for me on my older laptop running macOS Big Sur, but I just got my hands on a new Air (with the M1 processor, which may or may not be relevant) and I’m once again having this media keys issue. I also have some custom hot keys configured (which also used to work on my old laptop running Big Sur with the fixed build linked above) and they too fail to work as expected.

I’ve granted the app both _Accessibility_ and _Input Monitoring_ privileges in the Security & Privacy section of system settings, but these do not appear to have any effect.

**Comment by gfogus** [comment]
@agarzola Yep. Mine stopped working today also on a Mac system that was previously working after the fix.

**Comment by jgspratt** [comment]
Same here on macOS 10.15.7.  It suddenly just stopped working.  The fix from @martinduparc had been working for a few weeks but has now stopped working.

**Comment by HBBisenieks** [comment]
> Actually it's not the keys. I've spent a few hours tracking this down (linux) trying to figure out if there was any program hijacking the keys, but no.
> 
> If you use the same options on the menu they don't work either. Strange, as "Show" for example works as expected, as well as the rest of the menu (quit, help, etc).
> 
> I suspect the problem is not with GPMDP but some change from Google breaking things.

I can confirm that this is the case on MacOS 11.2 and 11.3. Media keys, tray icon controls, and touchbar controls all don't function, which is a drag and a half.

**Comment by agarzola** [comment]
This issue appears to have resolved itself! Can’t be sure when it was fixed, but I noticed it just now. Media keys, custom hot keys, and the menu bar submenu, all work for me on macOS Big Sur without having updated anything. I suggest others in this thread refresh the player and try their media keys and custom hot keys to confirm.

**Comment by alexander-narbut** [comment]
The only reason I use GPMDP is hotkeys and they're not working.
Tried to completely reinstall + remove the folder from %AppData%, nothing helps, Windows 10 19042.985.

**Comment by agzam** [comment]
Damn it. Has Google made some changes again that have broken the multimedia keys in the app? Again... sigh

**Comment by alexventuraio** [comment]
> Here is a macOS build with the updated ytmusic.js
> https://www.dropbox.com/s/8wg2lrlxjnwd9r0/Google%20Play%20Music%20Desktop%20Player.zip

Can anyone share this build, plase?
The file seems to be delted.

**Comment by bruno-brant** [comment]
> Can anyone share this build, plase?
> The file seems to be delted.

Hey @alexventuraio I suggest you look into YT Music Desktop (https://ytmdesktop.app) since it is still being maintained.

**Comment by juncas2205-debug** [comment]
> **Lista de verificación:**
> 
> * [x] 
>   [x] He comprobado que no hay problemas con contenidos similares o iguales.
>   **TAMBIÉN DEBERÍAS REVISAR LOS ASUNTOS CERRADOS**
>   [x] 
>   [x] He revisado las preguntas frecuentes ( https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/wiki/FAQ ) y la respuesta que busco no está allí.
>   [x] 
>   [x] Lo he comprobado dos veces y puedo reproducir el problema.
> 
> **Sistema operativo:** Windows
> 
> **Versión GPMDP:** 4.7.1
> 
> **Descripciones de problemas:** Los botones Fn + F5 / F6 / F7 (canción anterior, pausa, canción siguiente) dejaron de funcionar
> 
> **Pasos para reproducir:**
